### ISSUE TITLE
Add VMLA assembly formatters to tablegen

### ISSUE THREAD

**Issue by benvanik** [google/iree#1198]
All of the VMLA ops are currently using the generic printer/parser. To make the IR easier to read it'd be good to go through and use the tablegen assemblyFormat feature to make them look better.

**Comment by benvanik** [comment]
(replying here in case it gets lost in the closed PR)

We should be able to drop the ()'s from the op formatting too - this would match the core MLIR dialects that don't use the parens unless it's printing the generic form: https://mlir.llvm.org/docs/Dialects/Standard/#stdabsf-absfop

**Comment by phoenix-meadowlark** [comment]
I'll add a commit doing this to my next PR (#1624).

**Comment by phoenix-meadowlark** [comment]
# Buffer Manipulation Op Formatting
Here are some proposals for formatting the buffer manipulation ops. The first lines of each code block are the ops cleaned up a bit from standard form, but with all of the type information left intact and in a standard(ish) format for context.

### `buffer.const`
```mlir
%result = vmla.buffer.const %value : !vmla.buffer -> !iree.byte_buffer
%result = vmla.buffer.const %value : !iree.byte_buffer
```
Is there some way to indicate that `!iree.byte_buffer` refers to the return type? Maybe something like `: -> !iree.byte_buffer`, but less odd? As it is this op and `buffer.alloc` use the same syntax to refer to a single argument type on one hand and a result type on the other. It's obvious enough which refers to which if you know what the ops do, but do we want to make that assumption?
### `buffer.alloc`
```mlir
%result = vmla.buffer.alloc %byte_length : index -> !vmla.buffer
%result = vmla.buffer.alloc %byte_length : index
```
### `buffer.clone`
```mlir
%result = vmla.buffer.clone %src : !vmla.buffer -> !vmla.buffer
%result = vmla.buffer.clone %src
```
### `buffer.byte_length`
```mlir
%result = vmla.buffer.byte_length %value : !vmla.buffer -> index
%result = vmla.buffer.byte_length %value : index
```
Same question here as in `buffer.const`

### `buffer.view`
```mlir
%result = vmla.buffer.view %src, %byte_offset, %byte_length : (!vmla.buffer, index, index) -> !vmla.buffer
%result = vmla.buffer.view %src, %byte_offset, %byte_length : !vmla.buffer, index, index
%result = vmla.buffer.view %src(%byte_offset : index, %byte_length : index)
```
### `buffer.copy`
```mlir
vmla.buffer.copy %src, %src_byte_offset, out %dst, %dst_byte_offset, %dst_length : !vmla.buffer, index, !vmla.buffer, index, index
vmla.buffer.copy %src(%src_byte_offset : index), out %dst(%dst_byte_offset : index, %dst_length : index)
```
Aside: Why do copy and fill have `out` semantics but the others buffer ops don't?
### `buffer.fill`
```mlir
vmla.buffer.fill %value, out %dst : !vmla.buffer, !vmla.buffer
vmla.buffer.fill %value, out %dst
```
### `buffer.load.i32`
```mlir
%result = vmla.buffer.load.i32 %src, %byte_offset : (!vmla.buffer, index) -> i32
%result = vmla.buffer.load.i32 %src(%byte_offset : index) : i32
```

**Comment by benvanik** [comment]
it's not required to have types where the types are known - for example, vmla.buffer.fill always has matching arguments, so just `: !vmla.buffer` is enough (this is like `%c = std.add %a, %b : i32`)

here, it's known that all offsets are of type index (and they can't be anything else), so they can always be omitted

in general, the `: foo` suffix is used for the return type as well instead of `->` (that's used to denote results within a signature like `(operands) -> (results)`), so things like alloc would be:
```
%result = vmla.buffer.alloc %byte_length : !vmla.buffer
```

What I've gotten in the habit of doing to increase the readability is name args, so:
```
%result = vmla.buffer.alloc byte_length = %byte_length : !vmla.buffer
```
and
```
vmla.buffer.copy %src, src_byte_offset = %src_byte_offset, out %dst, dst_byte_offset = %dst_byte_offset, byte_length = %dst_length : !vmla.buffer
```
(as both %src and %dst must be the same and all offsets/lengths are index, so only !vmla.buffer indicating the types of %src and %dst is needed)

**Comment by GMNGeoffrey** [comment]
> ### `buffer.const`
> ```mlir
> %result = vmla.buffer.const %value : !vmla.buffer -> !iree.byte_buffer
> %result = vmla.buffer.const %value : !iree.byte_buffer
> ```
Make the byte_buffer type buildable and you can leave it off entirely.

> ### `buffer.alloc`
> ```mlir
> %result = vmla.buffer.alloc %byte_length : index -> !vmla.buffer
> %result = vmla.buffer.alloc %byte_length : index
> ```

index should be buildable and you can leave it off entirely.

> ### `buffer.load.i32`
> ```mlir
> %result = vmla.buffer.load.i32 %src, %byte_offset : (!vmla.buffer, index) -> i32
> %result = vmla.buffer.load.i32 %src(%byte_offset : index) : i32
> ```

Let's do `[]` for byte offsets, since they're indexing into something. This is consistent with MLIR standard ops. And again you can just drop index entirely. i32 should also be buildable

```mlir
%result = vmla.buffer.load.i32 %src[%byte_offset]
```

**Comment by GMNGeoffrey** [comment]
> so only !vmla.buffer indicating the types of %src and %dst is needed

We actually don't even need `!vmla.buffer` because it's a buildable type (now, at least)

**Comment by GMNGeoffrey** [comment]
> ```mlir
> vmla.buffer.copy %src, src_byte_offset = %src_byte_offset, out %dst, dst_byte_offset = %dst_byte_offset, byte_length = %dst_length : !vmla.buffer
> ```

I'd love to come up with a consistent way of saying "offset + length". `[]` for just an offset makes sense (like for load).

```mlir
%result = vmla.buffer.load.i32 %src[%byte_offset]
```

A bit trickier for copy and view. This looks weird (makes it seem 2D

```mlir
%result = vmla.buffer.view %src[%byte_offset, %byte_length]
```

It's too bad it's not a start/end offset, since that would be easier to do with something like one of these

```mlir
// Python-style syntax, but looks like a type
%result = vmla.buffer.view %src[%byte_offset:%byte_length]
// Maybe looks like a return type?
%result = vmla.buffer.view %src[%byte_offset -> %byte_length]
```

We could do a mixed with name arguments:

```mlir
%result = vmla.buffer.view %src[%byte_offset], byte_length = %byte_length
```

**Comment by phoenix-meadowlark** [comment]
> It's too bad it's not a start/end offset, since that would be easier to do with something like one of these
> ```mlir
> // Python-style syntax, but looks like a type
> %result = vmla.buffer.view %src[%start_offset:%end_offset]
> // Maybe looks like a return type?
> %result = vmla.buffer.view %src[%start_offset -> %end_offset]
> ```

We might still want to avoid it, but is there any ambiguity with the python-like syntax, or should it be inferable from `%end_offset` being a SSA value that it doesn't refer to a type?

**Comment by benvanik** [comment]
we don't want end offset, we want length (the common case is that the offset changes but the length does not, so we don't want to force end calculations that otherwise aren't needed)

**Comment by GMNGeoffrey** [comment]
> > It's too bad it's not a start/end offset, since that would be easier to do with something like one of these
> > ```mlir
> > // Python-style syntax, but looks like a type
> > %result = vmla.buffer.view %src[%start_offset:%end_offset]
> > // Maybe looks like a return type?
> > %result = vmla.buffer.view %src[%start_offset -> %end_offset]
> > ```
> 
> We might still want to avoid it, but is there any ambiguity with the python-like syntax, or should it be inferable from `%end_offset` being a SSA value that it doesn't refer to a type?

Yeah this would only work if the op were different. I'm just spitballing

**Comment by phoenix-meadowlark** [comment]
Okay, summarizing everything so far:

### `buffer.const`
```mlir
%result = vmla.buffer.const %value
```
### `buffer.alloc`
```mlir
%result = vmla.buffer.alloc byte_length = %byte_length
```
### `buffer.clone`
```mlir
%result = vmla.buffer.clone %src
```
### `buffer.byte_length`
```mlir
%result = vmla.buffer.byte_length %value
```
### `buffer.view`
```mlir
%result = vmla.buffer.view %src[%byte_offset], byte_length = %byte_length
```
### `buffer.copy`
```mlir
vmla.buffer.copy %src[src_byte_offset], out %dst[%dst_byte_offset], byte_length = %byte_length
```
### `buffer.fill`
```mlir
vmla.buffer.fill %value, out %dst
```
### `buffer.load.i32`
```mlir
%result = vmla.buffer.load.i32 %src[%byte_offset]
```

**Comment by benvanik** [comment]
I generally prefer to have the `: type` on things that return values to improve readability as then from the defining op of an ssa value you can tell the type, without having to guess, so:
buffer.clone %src : !vmla.buffer
buffer.byte_length %value : index
buffer.load.i32 ... : i32

(just nice for consistency)

**Comment by phoenix-meadowlark** [comment]
I updated my summary comment to avoid essentially duplicating the information again.

**Comment by GMNGeoffrey** [comment]
LGTM

**Comment by phoenix-meadowlark** [comment]
# Proposed Formatting for VMLA Shape/Structure Ops

### `vmla.copy`
```mlir
vmla.copy %src(%src_shape : !shapex.ranked_shape<[64]>)[%src_indicies_0, %src_indicies_1, ...]
          out %dst(%dst_shape : !shape.ranked_shape<[32]>)[%dst_indicies_0, %dst_indicies_1, ...]
          [%lengths_0, %lengths_1, ...] : i32
```

### `vmla.transpose` (for reference)
```mlir
vmla.transpose %src(%src_shape : !shapex.ranked_shape<[64,32]>),
               out %dst(%dst_shape : !shapex.ranked_shape<[32,64]>)
               {permutation = dense<[1, 0]> : tensor<2xi32>} : f32
```

### `vmla.reverse`
```mlir
vmla.reverse %src(%src_shape : !shapex.ranked_shape<[4, 8]>),
             out %dst(%dst_shape : !shapex.ranked_shape<[4, 8]>)
             {dimensions = dense<[1]> : tensor<i32>} : f32
```

### `vmla.pad`
```mlir
vmla.pad %src(%src_shape : !shapex.ranked_shape<[]>),
         %value(%value_shape : !shapex.ranked_shape<[]>),
         out %dst(%dst_shape) : !shapex.ranked_shape<[]>)
         {edge_padding_low = dense<> : tensor<i32>,
          edge_padding_high = dense<> : tensor<i32>,
          interior_padding = dense<> : tensor<i32>} : f32
```

### `vmla.broadcast`
```mlir
vmla.broadcast %src(%src_shape : !shapex.ranked_shape<[]>),
               out %dst(%dst_shape : !shapex.ranked_shape<[]>) : f32
```

### `vmla.tile`
```mlir
vmla.tile %src(%src_shape : !shapex.ranked_shape<[]>),
          out %dst(%dst_shape : !shapex.ranked_shape<[]>) : f32
```

### `vmla.gather`
```mlir
vmla.gather %src(%src_shape : !shapex.ranked_shape<[]>),
            %indices(%indices_shape : !shapex.ranked_shape<[]>),
            out %dst(%dst_shape : !shapex.ranked_shape<[]>)
            {dim = 1 : i64, batch_dims = 2 : i64} : f32
```

**Comment by phoenix-meadowlark** [comment]
# Proposed Formatting for VMLA Convolution/Reduction Ops

### `vmla.conv`
```mlir
vmla.conv %input(%input_shape : !shapex.ranked_shape<[]>) : f32,
          %filter(%filter_shape : !shape.ranked_shape<[]>) : f32,
          out %dst(%dst_shape : !shape.ranked_shape<[]>) : f32,
          {window_strides = dense<1> : vector<2xi32>,
           padding = dense<[1, 2, 2, 2]> : vector<4xi32>,
           lhs_dilation = dense<1> : vector<2xi32>,
           rhs_dilation = dense<1> : vector<2xi32>,
           feature_group_count = 1 : i32,
           batch_group_count = 1 : i32}
```

### `vmla.reduce.*`
```mlir
vmla.reduce.op %src(%src_shape : !shapex.ranked_shape<[4,8]>),
               %init(%init_shape : !shapex.ranked_shape<[]>),
               out %dst(%dst_shape : !shapex.ranked_shape<[4]>)
               {dimension = 1} : f32
```

### `vmla.pooling.*`
```mlir
vmla.pooling.op %src(%src_shape : !shapex.ranked_shape<[32,32]>),
                %init(%init_shape : !shapex.ranked_shape<[]>),
                out %dst(%dst_shape : !shapex.ranked_shape<[16,16]>)
                {window_dimensions = dense<[2,2]> : tensor<2xi32>,
                 window_strides = dense<[2,2]> : tensor<2xi32>} : f16
```

**Comment by phoenix-meadowlark** [comment]
# Proposed Formatting for VMLA Interface and Comparison Ops

### `vmla.interface.const`
```mlir
%result = vmla.interface.const %interface {offset = 3 : IREE_IndexAttr} : !vmla.buffer
```

### `vmla.interface.binding`
```mlir
%result = vmla.interface.binding %interface {binding = 0 : i32, set = 0 : i32} : !vmla.buffer
```

### `vmla.cmp`
```mlir
vmla.cmp %lhs, %rhs out %dst {predicate = 1 : i32} : f32
```

### `vmla.select`
```mlir
vmla.select %cond, %lhs, %rhs, out %dst : f32
```

**Comment by GMNGeoffrey** [comment]
> # Proposed Formatting for VMLA Shape/Structure Ops
> ### `vmla.copy`
> ```mlir
> vmla.copy %src(%src_shape : !shapex.ranked_shape<[64]>)[%src_indicies_0, %src_indicies_1, ...]
>           out %dst(%dst_shape : !shape.ranked_shape<[32]>)[%dst_indicies_0, %dst_indicies_1, ...]
>           [%lengths_0, %lengths_1, ...] : i32
> ```

Maybe `lengths =  [%lengths_0, %lengths_1, ...]`?

> ### `vmla.reverse`
> ```mlir
> vmla.reverse %src(%src_shape : !shapex.ranked_shape<[4, 8]>),
>              out %dst(%dst_shape : !shapex.ranked_shape<[4, 8]>)
>              {dimensions = dense<[1]> : tensor<i32>} : f32
> ```

Not really this issue, but reverse has the same input and output shape. Why does it have to be listed twice? @benvanik 

> ### `vmla.pad`
> ```mlir
> vmla.pad %src(%src_shape : !shapex.ranked_shape<[]>),
>          %value(%value_shape : !shapex.ranked_shape<[]>),
>          out %dst(%dst_shape) : !shapex.ranked_shape<[]>)
>          {edge_padding_low = dense<> : tensor<i32>,
>           edge_padding_high = dense<> : tensor<i32>,
>           interior_padding = dense<> : tensor<i32>} : f32
> ```

Assume this is a typo with an extra close-paren after `%dst_shape`

**Comment by GMNGeoffrey** [comment]
> ### `vmla.constant`
> ```mlir
> %result = "vmla.constant"() {value = dense<> : tensor<f32>} : () -> !vmla.buffer
> %result = vmla.constant value = dense<> : tensor<f32> : !vmla.buffer
> ```

The "value" name doesn't seem to useful to me. That's standard for constants and doesn't convey much extra information. Also the two colons is a bit hard to parse IMO. Maybe something like:

```mlir
%result = vmla.constant dense<> : tensor<f32> -> !vmla.buffer
```

Or throw the parens back in there

```mlir
%result = vmla.constant (dense<> : tensor<f32>) : !vmla.buffer
```

**Comment by GMNGeoffrey** [comment]
> ### `vmla.cmp`
> ```mlir
> vmla.cmp %lhs, %rhs out %dst {predicate = 1 : i32} : f32
> ```

Let's model this after http://google3/third_party/llvm/llvm-project/mlir/test/IR/core-ops.mlir?l=127&rcl=308312709

We should be able to get pretty printing to make the predicate have a name instead of an int. Either way should be able to leave off the type

```mlir
vmla.cmp %lhs, %rhs out %dst {predicate = 1} : f32
```

If we can make the predicate nicely printed:

```mlir
vmla.cmp "eq" %lhs, %rhs out %dst : f32
```

**Comment by phoenix-meadowlark** [comment]
> > ### `vmla.constant`
> > ```mlir
> > %result = "vmla.constant"() {value = dense<> : tensor<f32>} : () -> !vmla.buffer
> > %result = vmla.constant value = dense<> : tensor<f32> : !vmla.buffer
> > ```
> 
> The "value" name doesn't seem to useful to me. That's standard for constants and doesn't convey much extra information. Also the two colons is a bit hard to parse IMO. Maybe something like:
> 
> ```mlir
> %result = vmla.constant dense<> : tensor<f32> -> !vmla.buffer
> ```
> 
> Or throw the parens back in there
> 
> ```mlir
> %result = vmla.constant (dense<> : tensor<f32>) : !vmla.buffer
> ```

I think just eliding the `value =` is the cleanest. The only reason I included it was because I wasn't sure if it we wanted some kind of signal that it was coming from the attr-dict.

**Comment by phoenix-meadowlark** [comment]
> > ### `vmla.copy`
> > ```mlir
> > vmla.copy %src(%src_shape : !shapex.ranked_shape<[64]>)[%src_indicies_0, %src_indicies_1, ...]
> >           out %dst(%dst_shape : !shape.ranked_shape<[32]>)[%dst_indicies_0, %dst_indicies_1, ...]
> >           [%lengths_0, %lengths_1, ...] : i32
> > ```
> 
> Maybe `lengths = [%lengths_0, %lengths_1, ...]`?
> 
> > ### `vmla.reverse`
> > ```mlir
> > vmla.reverse %src(%src_shape : !shapex.ranked_shape<[4, 8]>),
> >              out %dst(%dst_shape : !shapex.ranked_shape<[4, 8]>)
> >              {dimensions = dense<[1]> : tensor<i32>} : f32
> > ```
> 
> Not really this issue, but reverse has the same input and output shape. Why does it have to be listed twice? @benvanik
> 
> > ### `vmla.pad`
> > ```mlir
> > vmla.pad %src(%src_shape : !shapex.ranked_shape<[]>),
> >          %value(%value_shape : !shapex.ranked_shape<[]>),
> >          out %dst(%dst_shape) : !shapex.ranked_shape<[]>)
> >          {edge_padding_low = dense<> : tensor<i32>,
> >           edge_padding_high = dense<> : tensor<i32>,
> >           interior_padding = dense<> : tensor<i32>} : f32
> > ```
> 
> Assume this is a typo with an extra close-paren after `%dst_shape`

I updated my original post to incorporate the first and the third comment here.

I was also wondering why the shape was listed twice. I'm assuming this op is essentially the same as `np.flip`?

**Comment by phoenix-meadowlark** [comment]
> > ### `vmla.cmp`
> > ```mlir
> > vmla.cmp %lhs, %rhs out %dst {predicate = 1 : i32} : f32
> > ```
> 
> Let's model this after http://google3/third_party/llvm/llvm-project/mlir/test/IR/core-ops.mlir?l=127&rcl=308312709
> 
> We should be able to get pretty printing to make the predicate have a name instead of an int. Either way should be able to leave off the type
> 
> ```mlir
> vmla.cmp %lhs, %rhs out %dst {predicate = 1} : f32
> ```
> 
> If we can make the predicate nicely printed:
> 
> ```mlir
> vmla.cmp "eq" %lhs, %rhs out %dst : f32
> ```

I'll look into how to implement this in another PR.
